portfolio/src/main/java/com/google/sps/servlets/ThumbsDownServlet.java

```diff
@@ -100,42 +102,58 @@ private void decrementPopularity(Entity commentEntity) {
         // Instantiate datastore
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-        // Get the previous thumbs down value.
-        long prevThumbsDown = (long) commentEntity.getProperty("thumbsdown");
-        // Get the previous popularity value.
-        long prevPopularity = (long) commentEntity.getProperty("popularity");
-
-        long newThumbsDown = prevThumbsDown + 1;
-        long newPopularity = prevPopularity - 1;
-        
-        // Update the thumbs down property to be the previous value plus one.
-        commentEntity.setProperty("thumbsdown", newThumbsDown);
-        // Update the popularity property to be the previous one minus one.
-        commentEntity.setProperty("popularity", newPopularity);
-
-        // Add the updated entity back in the datastore
-        datastore.put(commentEntity);
+        Transaction txn = datastore.beginTransaction();
```

Comment:
You need to get the comment entity inside the same transaction, otherwise it won't know what the initial state of the entity was. The basic flow of a transaction should be:

1. Begin Transaction
1. Get record(s)
1. Update values
1. Put record(s) back
1. End Transaction

See https://cloud.google.com/datastore/docs/concepts/transactions for an example
Important things to note here:

1. They use `txn.fetch` and `txn.put` to fetch and store data. This ensures those operations are done as part of the transaction.
1. They do both the fetch and the put as part of the same transaction.

Reply:
So I just realized we're using the old API here:
https://cloud.google.com/appengine/docs/standard/java/datastore/transactions

so the methods are slightly different `datastore.get` and `datastore.put`, but my initial comment still holds. You need to do both the get and the put inside the transaction and the put needs to have the `txn` as the first argument.

Reply:
I see, thank you! I just committed changes to address that issue.

Reply:
Nice! This transaction looks good, but there's one concern. Is it possible for a comment to change (due to actions from another thread) from being liked to unliked in-between when you read the unliked status and when you make changes to the comment? If so, this part would also need to be part of the transaction because there's a race condition.

Reply:
Yes, I think that the other thing that could happen at the same time was if a user tried to delete a comment while the comment was being up voted or down voted.